docs/proposals/20220206-quota.md

```diff
@@ -0,0 +1,257 @@
+* Given quota admission control is a proprietary implementation for different cloud environments. Providing an universal solution is not pratical. This is why in this document we propose to build a plugin to let users to easily integrate their quota validation logics into pod-syncer.
+
+### Goals
+
+* Propose a quota validation plugin for user to integrate proprietary quota management functions into pod-syncer.
+* Define per tenant locking mechanism in pod-syncer to preserve parallel processing of tenants’ requests and prevent race condition while checking quota usage.
+* Give an example of quota validation plugin registration and initialization.
+
+### Non-Goals
+
+* Queuing mechanism used for managing pods that exceeds quota limit.
+* Define design and implementations for quota provision, quota validation and quota usage tracking.
+
+## Proposal
+
+### User Stories and Requirements
+
+* Enforce tenant overall quota usage in super cluster but allow tenant freely creating namespace, and managing separate quota in virtual cluster.
+* Enable cross namespace quota admission control in super cluster to reduce resource fragementation.
+* Allow cloud admin to define own quota validation implementation and easy integration into pod-syncer
+* Handle tenant’s resource requests in parallel and minimize interferences of large batch requests from noisy tenant.
+
+### Implementation Details
+
+Given virtual cluster syncer a unique place that aggregates and processes all tenants’ requests through fair queuing,  we think the best place to validate quota is in pod-syncer. This is because:
+
+* pod-syncer has all knowledge of namespace/tenant mapping, resource usage status and resource request details.
+* All kind resource requests, e.g. through job creation, deployment, or from replica-set, all end up being processed in pod-syncer, which reduce the capturing and converting different resource requests.
+* pod-syncer has all privilege of controlling resources in super cluster, able to hold tenant’s pod in pending state, and re-queue the pending pods periodically.
+* the syncer has unique queuing mechanism that allows locking of one tenant in a short period of time without interfering other tenants.
+
+
+Following diagram depicts how the validation plugin works together with pod-syncer.
+![Resource Admission Control Plugin](images/quota/plugin.png)
+When pod syncer reconciler receives pod creation request:
+
+* it first checks if quota validation flag is on, if not, it goes ahead with normal pod creation procedure.
+* if quota validation flag is on, pod-syncer searches for quota validation plugin that is dynamically registered at syncer starting time, and call the quota validation plugin
+* the pod-syncer first locks tenant, then get current tenant quota usage, compares available quota with requested quota from pod, if there is enough quota, pod syncer will creates the pod in super cluster and removes the per tenant lock.
```

Comment:
We should explain by how this will be implemented upstream, not by how a validator should be written since that could be anything.

```suggestion
* the pod-syncer first locks tenant, then calls the quota validation plugin and if the validation is successful the pod syncer will create the pod in super cluster and removes the per tenant lock.
```